xla/mlir/backends/openxla/conversion/de_bufferization.h

```diff
@@ -0,0 +1,108 @@
+// key difference is that at LMHLO level program operates on buffers (memrefs),
+// and at IREEInput level it uses value semantics and tensors.
+//
+// We rely on in-place semantics of IREEInput operations (tied operands) to
+// re-write operations writing to buffers to operations updating tensors.
+//
+// Example:
+//
+//   func @main(%arg0: memref<f32>, %arg1: memref<f32>) {
+//     lmhlo.fusion {
+//       %0 = bufferization.to_tensor %arg0 : memref<f32>
+//       %0 = bufferization.to_tensor %arg1 : memref<f32>
+//       %2 = mhlo.add %0, %1: tensor<f32>
+//       memref.tensor_store %2, %arg1 : memref<f32>
+//     }
+//     "some.consumer"(%arg1) : (memref<f32>) -> ()
+//   }
+//
+// In this example `%arg0` is a read only buffer, and `%arg1` is a read-write
+// buffer.
+//
+//  func @main(%arg0: tensor<f32>, %arg1: tensor<f32>) {
+//     %0 = iree_input.dispatch @add(%arg0, %arg1)
+//       : (tensor<f32>, tensor<f32>) -> %arg1
+//     "some.consumer"(%0) : (tensor<f32>) -> ()
+//   }
+//
+// We use `DeBufferization` to track the mapping from a memref to the last
+// tensor produced by an operation that wrote into a memref. In the example
+// above instead of passing `%arg1` to the consumer, we pass the last tensor
+// that shares the storage with `%arg1`. After lowering this representation to
+// IREEs HAL dialect, it's guaranteed that XLA program will read/write from/to
+// exactly the same buffer slices as its original version.
+//
+// In XLA all input arguments do not alias, so we don't need any buffer aliasing
+// analysis, and we can safely rely on memref.view operation offsets and sizes.
+//
+// Conversion implementation is a bit more complicated because we have to handle
+// memref.view and memref.reinterpret_cast operations, but all conversions
+// conceptually are doing the same transformation as in example above.
+struct DeBufferization {
+  // Mapping block block arguments to memref views constructed from them. We'll
+  // need it at the very end to tie all inplace updates to the optimization
+  // barrier to prevent dead code elimination.
+  llvm::DenseMap<mlir::BlockArgument,
+                 llvm::SmallVector<mlir::TypedValue<mlir::MemRefType>>>
+      imported;
+
+  // Mapping from the memref view to the last tensor that is tied to the same
+  // underlying storage. We use this mapping to thread inplace tensor updates
+  // through all operations in the compiled function.
+  llvm::DenseMap<mlir::Block *,
+                 llvm::DenseMap<mlir::TypedValue<mlir::MemRefType>,
+                                mlir::TypedValue<mlir::TensorType>>>
+      remapped;
+};
+
+// We only pass around tensors constructed from a row major memrefs because
+// currently IREE buffer view can't represent a strided layout. As a short term
+// solution the plan is to pass tensor layout as a side data structure, but
+// longer term we'll need to add tensor/buffer layouts to OpenXLA/IREE.
```

Comment:
Could open an issue and get Ben's input on this. Looks like `flow.dispatch.tensor.load` has strides (https://openxla.github.io/iree/reference/mlir-dialects/Flow/#flowdispatchtensorload-flowdispatchtensorloadop), but I don't see strides in the Input or HAL dialects